### PR TITLE
apk: add package

### PR DESCRIPTION
**Pull request by aparcar**
APK (Alpine Package Keeper) is the package manager of Alpine Linux and has multiple advantages over OPKG. While Alpine uses APK version 2, this commit adds version 3 with a heavily optimised database structure and additional feature making it suitable for OpenWrt.

This commit will be followed by many more to add APK build capabilities to the OpenWrt build system, firstly enabling side by side builds of APK and OPKG packages, later replacing OPKG entirely.

### CODE REVIEW

File: package/system/apk/Makefile [hunk 1]
```diff
@@ -0,0 +1,66 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=apk
+PKG_RELEASE:=1
+
+PKG_SOURCE_DATE:=2024-03-16
+PKG_SOURCE_VERSION:=f2f10d7e851ed3ff2a239c89c1a6a3612adbd63f
+PKG_HASH:=964bbc71462630a5418a20ccd3e725eb0b065a6232d4aee16db745d95a04c1ff
+
+PKG_VERSION:=3.0.0_pre0
```
Comment (dangowrt) [inline package/system/apk/Makefile:10]: We will need incremental preXXXXX versions because we will probably still bump to git HEAD a couple of times before the official APK 3.0.0 release.

File: package/system/apk/Makefile [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	-Ddocs=disabled \
+	-Dzstd=False
+MESON_ARGS += \
+	-Dlua_version=5.1 \
+	-Dcompressed-help=false \
+	-Ddocs=disabled \
+	-Dzstd=False
+	
+HOST_LDFLAGS += \
+	-Wl,-rpath $(STAGING_DIR_HOST)/lib \
```
Comment (neheb) [inline package/system/apk/Makefile:49]: NACK. host packages should use static libraries. See https://github.com/openwrt/openwrt/pull/14694

File: package/system/apk/Makefile [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	-lpthread
+
+TARGET_CFLAGS += $(FPIC)
```
Comment (neheb) [inline package/system/apk/Makefile:52]: meson should be adding fpic by default.

Comment (aparcar) [reply package/system/apk/Makefile]: Removed

File: package/system/apk/Makefile (hunk 2 above)
Comment (aparcar) [reply package/system/apk/Makefile]: Removed

File: package/system/apk/Makefile (hunk 1 above)
Comment (aparcar) [reply package/system/apk/Makefile:10]: Can't we just use PKG_RELEASE for that?

File: package/system/apk/Makefile [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	-Dcompressed-help=false \
+	-Ddocs=disabled \
+	-Dzstd=False
+MESON_ARGS += \
+	-Dlua_version=5.1 \
+	-Dcompressed-help=false \
+	-Ddocs=disabled \
+	-Dzstd=False
+	
+HOST_LDFLAGS += \
```
Comment (neheb) [inline package/system/apk/Makefile:48]: why is this needed?

File: package/system/apk/Makefile [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/host-build.mk
+include $(INCLUDE_DIR)/meson.mk
+
+define Package/apk
+  SECTION:=base
+  CATEGORY:=Base system
+  TITLE:=apk package manager
+  DEPENDS:=+zlib +libopenssl +libpthread @!arc
```
Comment (neheb) [inline package/system/apk/Makefile:33]: lpthread dependency is noop with latest musl and glibc.

**Comment by aparcar** [comment]
@neheb could you please help me with this patch? https://gitlab.alpinelinux.org/alpine/apk-tools/-/merge_requests/160

File: package/system/apk/Makefile (hunk 1 above)
Comment (aparcar) [reply package/system/apk/Makefile]: I added a patch so that the Git date is added to the binary itself

**Comment by stangri** [comment]
@aparcar is the long-term plan to keep the openssl dependency or make variants of this package with dependencies on other SSL libraries?

**Comment by aparcar** [comment]
> @aparcar is the long-term plan to keep the openssl dependency or make variants of this package with dependencies on other SSL libraries?

I think there are efforts in making it more flexible[1] but it will certainly take more time. I'd be very happy to see more people jumping on the project and fix the missing bits required to make it fully integrated in the OpenWrt ecosystem. More prominently is the switch from libfetch to uclient.

[1]: https://gitlab.alpinelinux.org/alpine/apk-tools/-/merge_requests/59

**Comment by liushuai42** [comment]
Is it advisable to choose a notable name that conflicts with a well-known Android Package Kit?

**Comment by neheb** [comment]
Upstream knows: https://gitlab.alpinelinux.org/alpine/apk-tools/-/issues/10769

**Comment by aparcar** [comment]
This works and runs fine on my devices. I'd merge it as soon as #15092 makes it through.

**Comment by aparcar** [comment]
Merged.
